src/psyclone/tests/dynamo0p3_builtins_test.py

```diff
@@ -2583,12 +2638,15 @@ def test_scalar_int_builtin_error(monkeypatch):
         _, invoke_info = parse(
             os.path.join(BASE_PATH, "16.2_integer_scalar_sum.f90"),
             api="dynamo0.3", distributed_memory=dist_mem)
-        with pytest.raises(ParseError) as excinfo:
-            _ = PSyFactory("dynamo0.3",
-                           distributed_memory=dist_mem).create(invoke_info)
-        assert ("an argument to a built-in kernel must be one of ['gh_field', "
-                "'gh_real'] but kernel " + test_builtin_name.lower() + " has "
-                "an argument of type gh_integer" in str(excinfo))
+        # Attempt to generate the code
+        psy = PSyFactory("dynamo0.3",
+                         distributed_memory=dist_mem).create(invoke_info)
+        with pytest.raises(GenerationError) as excinfo:
```

Comment:
I think the problem is that you've moved the check for the raising of `GenerationError` to be around the call to `psy.gen` whereas it still needs to be around the call to `PSyFactory(blah).create(blah)` - see how it was before.

Reply:
Thanks! I thought that GenerationError was a property of psy.gen, hence moving the call. I tried several things and I'm afraid the test still doesn't work (list below).

1. Moved the check around PSyFactory as it was before and removed call to psy before the check.

        with pytest.raises(GenerationError) as excinfo:
            _ = PSyFactory("dynamo0.3",
                           distributed_memory=dist_mem).create(invoke_info)
        assert ("Integer reductions are not currently supported "
                "by the LFRic infrastructure. Error found in "
                "Kernel '" + test_builtin_name.lower() + "', argument 'isum'"
                in str(excinfo.value)

returns failure:

E               Failed: DID NOT RAISE <class 'psyclone.psyGen.GenerationError'>

The same happens if I move the call to psy within the check.

2. Put back the psy = PSyFactory(... call before the check. Now it fails with 

E                       GenerationError: "Generation Error: Integer reductions are not currently supported by the LFRic infrastructure. Error found in Kernel 'x_innerproduct_y', argument 'isum'"

3. Intentionally put the wrong info to check for ParseError (as it was before). It fails with GenerationError again.

Reply:
I've just had a look and realised that we fail to raise the error if distributed memory is not enabled!
We'll need to add a check for written, integer arguments earlier on - there must be a place where we validate the kernel arguments?

Reply:
Got it now, thank you! I modified the test to call `PSyFactory` only  `if dist_mem` and I everything is now fine (I tested it for `True`). I am not sure what error to raise if `dist_mem` is not enabled, as the call to `PSyFactory` for this test only makes sense when it is enabled.

Regarding the checks for written, integer arguments you mentioned, I think a number of tests will need to be updated anyway (please see the updated description of this PR).